### PR TITLE
Fix #3581: Create jupyterhub user given email that may not be Sirepo …

### LINKED ISSUE #3581: pkcli.jupyterhublogin.create_user add display_name
I realized that we'll need a display name to be added. Please add a non-optional parameter to create_user.

### PR DESCRIPTION
**Pull request by e-carlin**
…user

@robnagler take a close look at this one. I'm not 100% certain I got everything right regarding the edge cases of registration states a user may be in. I also had to refactor a small amount of auth code.

### CODE REVIEW

File: sirepo/auth/__init__.py [hunk 1]
```diff
@@ -495,19 +501,21 @@ def user_name():
     )
 
 
-def user_registration(uid):
+def user_registration(uid, display_name=None):
```
Comment (robnagler) [inline sirepo/auth/__init__.py:504]: Remove =None

File: sirepo/pkcli/jupyterhublogin.py [hunk 2]
```diff
@@ -25,14 +33,54 @@ def create_user(email):
...
     import sirepo.template
 
+    def maybe_create_sirepo_user(module, email, display_name):
+        u = module.unchecked_user_by_user_name(email)
+        if u:
+            # Fully registered email user
+            r = sirepo.auth_db.UserRegistration.search_by(uid=u)
+            assert r.display_name, \
+                f'uid={u} authorized AuthEmailUser record but no UserRegistration.display_name'
+            if not r.display_name == display_name:
```
Comment (robnagler) [inline sirepo/pkcli/jupyterhublogin.py:43]: Don't change the name, just leave it. The purpose is not to ensure a name. Let the person's registration override what we do

File: sirepo/pkcli/jupyterhublogin.py [hunk 3]
```diff
@@ -25,14 +33,54 @@ def create_user(email):
...
+                )
+                r.display_name = display_name
+                r.save()
+            return u
+        m = module.AuthEmailUser.search_by(unverified_email=email)
+        if m:
+            # Email user that needs to complete registration (no display_name but have unverified_email)
+            assert sirepo.auth.need_complete_registration(m), \
+                'email={email} has no display_name but does not need to complete registration'
+            pkcli.command_error(
```
Comment (robnagler) [inline sirepo/pkcli/jupyterhublogin.py:58]: good

**Review by robnagler — CHANGES_REQUESTED** [review]

File: sirepo/auth/__init__.py (hunk 1 above)
Comment (e-carlin) [reply sirepo/auth/__init__.py:504]: Why? In the non-pkcli case we don't have a display name when this is called. Should I just pass in None in that case?

File: sirepo/pkcli/jupyterhublogin.py (hunk 2 above)
Comment (e-carlin) [reply sirepo/pkcli/jupyterhublogin.py]: b5e8a8431

**Comment by e-carlin** [comment]
@robnagler comments addressed. I left [one question](https://github.com/radiasoft/sirepo/pull/3588#discussion_r634485167).

File: sirepo/auth/__init__.py (hunk 1 above)
Comment (robnagler) [reply sirepo/auth/__init__.py:504]: I was confused. I thought this was only for the pkcli. That's fine. leave as is.

**Pull request merged**